File: typescript/packages/otel/src/exporter.ts
import {
  ConfigError,
  type Exporter,
  type SkippedBranch,
  type Store,
  type SyncError,
  type SyncReport,
  VERSION,
} from "@threads/core";
import { err, ok, type Result } from "@threads/core/host";
import {
  type ChainEvent,
  type ChangedBranch,
  Feed,
  telemetryBinding,
} from "@threads/core/internal/feed";
import { Backoff } from "./backoff";
import type { Config } from "./env";
import { lossSpans } from "./losses";
import { batches, body, MAX_BATCH } from "./otlp";
import { post } from "./send";
import type { Span } from "./span";
import { type Branch, spans } from "./spans";

// Exporter.sync() (spec/otel/README.md, "The cursor, sync and losses"): every changed branch's
// newly closed spans, posted in batches; a branch's cursor moves only after the collector
// accepted its spans, so a crash re-sends with the same ids and never drops one.

type Ready = {
  readonly branch: ChangedBranch;
  readonly head: number;
  readonly spans: readonly Span[];
};

export class OtelExporter implements Exporter {
  readonly #config: Config;
  readonly #store: Store | undefined;
  readonly #observer: string;
  readonly #content: boolean;
  readonly #backoff: Backoff = new Backoff();
  #queue: Promise<unknown> = Promise.resolve();

  constructor(
    config: Config,
    store: Store | undefined,
    observer: string,
    content: boolean,
  ) {
    this.#config = config;
    this.#store = store;
    this.#observer = observer;
    this.#content = content;
  }

  /** One sync at a time: a second call waits for the first. */
  sync(): Promise<Result<SyncReport, SyncError>> {
    const run = (async (): Promise<Result<SyncReport, SyncError>> => {
      await this.#queue;
      return await this.#sync();
    })();
    this.#queue = settled(run);
    return run;
  }

  async #sync(): Promise<Result<SyncReport, SyncError>> {
    const binding = telemetryBinding(this);
    const signal = binding?.signal;
    const store = this.#store ?? binding?.store;
    if (store === undefined)
      throw new ConfigError(
        "invalid_config",
        "otel(): no store to export: pass otel({store}), or pass the exporter to host({telemetry})",
      );
    const feed = await Feed.open(store, this.#observer);
    await feed.register();
    const changed = await feed.changed();
    if (!changed.ok)
      throw new Error(
        `the store's branch rows are corrupt: ${changed.error.message}`,
      );
    const skipped: SkippedBranch[] = [];
    const ready = await this.#read(feed, changed.value, skipped, signal);
    if (signal?.aborted) return err(stopped());
    await feed.checkpoint(
      ready
        .filter((r) => r.spans.length === 0)
        .map((r) => ({ branch_id: r.branch.branch_id, seq: r.head })),
    );
    const sent = await this.#send(feed, ready, signal);
    if (!sent.ok) return sent;
    const lost = await this.#losses(feed, signal);
    if (!lost.ok) return lost;
    return ok({ spans: sent.value, possiblyLostEvents: lost.value, skipped });
  }

  /**
   * Each changed branch's chain and newly closed spans; one that doesn't read is skipped. It
   * yields to the event loop after each branch, so re-deriving many long threads never holds
   * the host's other work for more than one branch at a time.
   */
  async #read(
    feed: Feed,
    changed: readonly ChangedBranch[],
    skipped: SkippedBranch[],
    signal: AbortSignal | undefined,
  ): Promise<Ready[]> {
    const chains = new Map<string, readonly ChainEvent[] | undefined>();
    const ready: Ready[] = [];
    const now = Date.now();
    for (const branch of changed) {
      await nextTask();
      if (signal?.aborted) break;
      if (this.#backoff.waiting(branch.branch_id, branch.head_seq, now))
        continue;
      const read = await feed.chain(branch.branch_id);
      if (!read.ok) {
        this.#backoff.failed(branch.branch_id, branch.head_seq, now);
        const { branch_id, thread_id, head_seq } = branch;
        skipped.push({ branch_id, thread_id, head_seq, code: read.error.code });
        continue;
      }
      this.#backoff.cleared(branch.branch_id);
      const chain = read.value.events;
      const head = chain.at(-1)?.event.seq ?? branch.head_seq;
      const all = await spansOf(
        feed,
        {
          tenant: branch.tenant_id,
          branchId: branch.branch_id,
          chain,
          content: this.#content,
        },
        chains,
      );
      const fresh = all.filter(
        (s) => s.closeSeq > branch.cursor && s.closeSeq <= head,
      );
      ready.push({ branch, head, spans: fresh });
    }
    return ready;
  }

  /** Posts the spans in batches; after each 2xx, each branch in it moves as far as it is sent. */
  async #send(
    feed: Feed,
    ready: readonly Ready[],
    signal: AbortSignal | undefined,
  ): Promise<Result<number, SyncError>> {
    const byId = new Map<string, Ready>(
      ready.map((r) => [r.branch.branch_id, r]),
    );
    const left = new Map<string, number>(
      ready.map((r) => [r.branch.branch_id, r.spans.length]),
    );
    let sent = 0;
    for (const batch of batches(ready.flatMap((r) => r.spans))) {
      const posted = await post(
        this.#config,
        body(batch, this.#config.resource, VERSION),
        signal,
      );
      if (!posted.ok) return err(posted.error);
      // Stopped after the 2xx: the cursor stays, so these are sent again (never lost).
      if (signal?.aborted) return err(stopped());
      sent += batch.length;
      const through = new Map<string, number>();
      for (const s of batch) {
        left.set(s.branchId, (left.get(s.branchId) ?? 0) - 1);
        through.set(
          s.branchId,
          Math.max(through.get(s.branchId) ?? 0, s.closeSeq),
        );
      }
      await feed.checkpoint(
        [...through].flatMap(([id, seq]) => {
          const r = byId.get(id);
          if (r === undefined) return [];
          return [
            {
              branch_id: r.branch.branch_id,
              seq: left.get(id) === 0 ? r.head : seq,
            },
          ];
        }),
      );
    }
    return ok(sent);
  }

  /** The unreported deletion losses, as possibly_lost spans after the branch batches. */
  async #losses(
    feed: Feed,
    signal: AbortSignal | undefined,
  ): Promise<Result<number, SyncError>> {
    const rows = await feed.unreportedLosses();
    if (!rows.ok)
      throw new Error(
        `the store's loss rows are corrupt: ${rows.error.message}`,
      );
    let lost = 0;
    for (let at = 0; at < rows.value.length; at += MAX_BATCH) {
      const chunk = rows.value.slice(at, at + MAX_BATCH);
      const posted = await post(
        this.#config,
        body(lossSpans(this.#observer, chunk), this.#config.resource, VERSION),
        signal,
      );
      if (!posted.ok) return err(posted.error);
      if (signal?.aborted) return err(stopped());
      await feed.markReported(chunk);
      lost += chunk.reduce((n, r) => n + r.unchecked_events, 0);
    }
    return ok(lost);
  }
}

/**
 * `spans` over chains read from the feed: `spans` is synchronous and the store is not, so it
 * runs until it asks for no chain it doesn't have (each parent it walks names the next one).
 * `chains` caches across branches; an unreadable chain is cached as undefined.
 */
async function spansOf(
  feed: Feed,
  branch: Branch,
  chains: Map<string, readonly ChainEvent[] | undefined>,
): Promise<readonly Span[]> {
  for (;;) {
    const missing = new Set<string>();
    const found = spans(branch, (id) => {
      if (!chains.has(id)) missing.add(id);
      return chains.get(id);
    });
    if (missing.size === 0) return found;
    for (const id of missing) {
      const read = await feed.chain(id);
      chains.set(id, read.ok ? read.value.events : undefined);
    }
  }
}

function stopped(): SyncError {
  return {
    code: "collector_unavailable",
    message:
      "the host stopped: nothing more is sent, and the next start sends it again",
  };
}

/** A macrotask turn: timers, I/O and other work run before the next branch is read. */
function nextTask(): Promise<void> {
  const next = Promise.withResolvers<void>();
  setTimeout(next.resolve, 0);
  return next.promise;
}

/** Waits for `work` to end, whatever its outcome, so the next sync runs after it. */
async function settled(work: Promise<unknown>): Promise<void> {
  try {
    await work;
  } catch {
    // The caller of that sync sees its error.
  }
}

File: python/src/threads/otel/exporter.py
"""`Exporter.sync()` (spec/otel/README.md, "The cursor, sync and losses"): every changed
branch's newly closed spans, posted in batches; a branch's cursor moves only after the collector
accepted its spans, so a crash re-sends with the same ids and never drops one."""

import asyncio
import time
from collections import Counter
from dataclasses import dataclass

from pydantic.experimental.missing_sentinel import MISSING

from threads import VERSION
from threads.agents.config import ConfigError
from threads.agents.store import Store, now_ms, open_store
from threads.log import BranchId, ThreadStartedEvent
from threads.otel.backoff import Backoff
from threads.otel.env import Config
from threads.otel.losses import loss_spans
from threads.otel.otlp import MAX_BATCH, batches, body
from threads.otel.send import post
from threads.otel.span import Span
from threads.otel.spans import Branch, spans
from threads.result import Err, Ok
from threads.store._feed import ChangedBranch, Checkpoint, Feed
from threads.store.verify import VerifiedLog
from threads.telemetry import SkippedBranch, SyncError, SyncReport, bound_store


@dataclass(frozen=True, slots=True)
class _Ready:
    branch: ChangedBranch
    head: int
    spans: tuple[Span, ...]


class OtelExporter:
    """spec/api.json `Exporter`, made by `otel()`."""

    def __init__(
        self, config: Config, store: Store | None, observer: str, *, content: bool
    ) -> None:
        self._config = config
        self._store = store
        self._observer = observer
        self._content = content
        self._backoff = Backoff()
        self._lock = asyncio.Lock()

    async def sync(self) -> Ok[SyncReport] | Err[SyncError]:
        """One sync at a time: a second call waits for the first."""
        async with self._lock:
            return await self._sync()

    async def _sync(self) -> Ok[SyncReport] | Err[SyncError]:
        store = self._store if self._store is not None else bound_store(self)
        if store is None:
            raise ConfigError(
                "invalid_config",
                "otel(): no store to export: pass otel(store=...), or pass the exporter to"
                " host(telemetry=...)",
            )
        feed = (await open_store(store)).feed(self._observer, now_ms)
        await feed.register()
        skipped: list[SkippedBranch] = []
        ready = await self._read(feed, await feed.changed(), skipped)
        await feed.checkpoint(
            [Checkpoint(r.branch.branch_id, r.head) for r in ready if not r.spans]
        )
        sent = await self._send(feed, ready)
        if isinstance(sent, Err):
            return sent
        lost = await self._losses(feed)
        if isinstance(lost, Err):
            return lost
        return Ok(SyncReport(sent.value, lost.value, tuple(skipped)))

    async def _read(
        self, feed: Feed, changed: list[ChangedBranch], skipped: list[SkippedBranch]
    ) -> list[_Ready]:
        """Each changed branch's chain and newly closed spans; one that doesn't read is
        skipped."""
        chains: dict[str, VerifiedLog | None] = {}
        """Every chain read this sync, for a child's parents."""
        ready: list[_Ready] = []
        now = time.monotonic_ns() // 1_000_000
        for branch in changed:
            if self._backoff.waiting(branch.branch_id, branch.head_seq, now):
                continue
            read = await feed.chain(branch.branch_id)
            if isinstance(read, Err):
                self._backoff.failed(branch.branch_id, branch.head_seq, now)
                skipped.append(
                    SkippedBranch(
                        branch.branch_id, branch.thread_id, read.error.code, branch.head_seq
                    )
                )
                continue
            self._backoff.cleared(branch.branch_id)
            chains[branch.branch_id] = read.value
            ready.append(await self._fresh(feed, branch, read.value, chains))
            # Deriving a long chain is CPU work on the loop: let the host's other work run
            # between branches.
            await asyncio.sleep(0)
        return ready

    async def _fresh(
        self,
        feed: Feed,
        branch: ChangedBranch,
        log: VerifiedLog,
        chains: dict[str, VerifiedLog | None],
    ) -> _Ready:
        """The branch's spans that closed past its cursor, parents read on demand."""
        wanted = _parent_of(log)
        while wanted is not None and wanted not in chains:
            read = await feed.chain(wanted)
            chains[wanted] = read.value if isinstance(read, Ok) else None
            wanted = None if isinstance(read, Err) else _parent_of(read.value)
        head = log.head.seq
        found = spans(Branch(branch.tenant_id, branch.branch_id, log, self._content), chains.get)
        fresh = tuple(s for s in found if branch.cursor < s.close_seq <= head)
        return _Ready(branch, head, fresh)

    async def _send(self, feed: Feed, ready: list[_Ready]) -> Ok[int] | Err[SyncError]:
        """Posts the spans in batches; after each 2xx, each branch in it moves as far as it is
        sent."""
        by_id: dict[str, _Ready] = {r.branch.branch_id: r for r in ready}
        left: Counter[str] = Counter({b: len(r.spans) for b, r in by_id.items()})
        sent = 0
        for batch in batches([s for r in ready for s in r.spans]):
            posted = await post(self._config, body(batch, self._config.resource, VERSION))
            if isinstance(posted, Err):
                return posted
            sent += len(batch)
            through: dict[str, int] = {}
            for s in batch:
                left[s.branch_id] -= 1
                through[s.branch_id] = max(through.get(s.branch_id, 0), s.close_seq)
            await feed.checkpoint(
                [
                    Checkpoint(by_id[b].branch.branch_id, by_id[b].head if left[b] == 0 else seq)
                    for b, seq in through.items()
                ]
            )
        return Ok(sent)

    async def _losses(self, feed: Feed) -> Ok[int] | Err[SyncError]:
        """The unreported deletion losses, as possibly_lost spans after the branch batches."""
        rows = await feed.unreported_losses()
        lost = 0
        for at in range(0, len(rows), MAX_BATCH):
            chunk = rows[at : at + MAX_BATCH]
            spans_ = loss_spans(self._observer, chunk)
            posted = await post(self._config, body(spans_, self._config.resource, VERSION))
            if isinstance(posted, Err):
                return posted
            await feed.mark_reported(chunk)
            lost += sum(r.unchecked_events for r in chunk)
        return Ok(lost)


def _parent_of(log: VerifiedLog) -> BranchId | None:
    """The branch a child thread's first turn is parented through, if it is a child."""
    for segment in log.segments[:1]:
        for event, _ in segment.events[:1]:
            if isinstance(event, ThreadStartedEvent) and event.data.parent is not MISSING:
                return event.data.parent.branch_id
    return None
